asdf/_block/io.py

```diff
@@ -0,0 +1,521 @@
+        else:
+            msg = "write_block received offset for non-seekable file"
+            raise ValueError(msg)
+    fd.write(struct.pack(b">H", len(header)))
+    fd.write(header)
+    if buff is None:  # data is uncompressed
+        fd.write_array(data)
+    else:
+        fd.write(buff.getvalue())
+    fd.fast_forward(padding_bytes)
+    return BLOCK_HEADER.unpack(header)
+
+
+def _candidate_offsets(min_offset, max_offset, block_size):
+    offset = (max_offset // block_size) * block_size
+    if offset == max_offset:
+        offset -= block_size
+    while offset > min_offset:
+        yield offset
+        offset -= block_size
+    if offset <= min_offset:
+        yield min_offset
+
+
+def find_block_index(fd, min_offset=None, max_offset=None):
+    """
+    Find the location of an ASDF block index within a seekable file.
+
+    Searching will begin at the end of the file (or max_offset
+    if it is provided).
+
+    Parameters
+    ----------
+
+    fd : file or generic_io.GenericIO
+        A seekable file that will be searched to try and find
+        the start of an ASDF block index within the file.
+
+    min_offset : int, optional
+        The minimum search offset. A block index will not be
+        found before this point.
+
+    max_offset : int, optional
+        The maximum search offset. A block index will not be
+        found after this point.
+
+    Returns
+    -------
+
+    offset : int or None
+        Index of start of ASDF block index. This is the location of the
+        ASDF block index header.
+
+    """
+    if min_offset is None:
+        min_offset = fd.tell()
+    if max_offset is None:
+        fd.seek(0, os.SEEK_END)
+        max_offset = fd.tell()
+    block_size = fd.block_size
+    block_index_offset = None
+    buff = b""
+    pattern = constants.INDEX_HEADER
+    for offset in _candidate_offsets(min_offset, max_offset, block_size):
+        fd.seek(offset)
+        buff = fd.read(block_size) + buff
+        index = buff.find(pattern)
+        if index != -1:
+            block_index_offset = offset + index
+            if block_index_offset >= max_offset:
+                return None
+            break
+        buff = buff[: len(pattern)]
+    return block_index_offset
+
+
+def read_block_index(fd, offset=None):
+    """
+    Read an ASDF block index from a file.
+
+    Parameters
+    ----------
+
+    fd : file or generic_io.GenericIO
+        File to read the block index from.
+
+    offset : int, optional
+        Offset within the file where the block index starts
+        (the start of the ASDF block index header). If not provided
+        reading will start at the current position of the file
+        pointer. See `find_block_index` to locate the block
+        index prior to calling this function.
+
+    Returns
+    -------
+
+    block_index : list of ints
+        A list of ASDF block offsets read and parsed from the
+        block index.
+
+    Raises
+    ------
+    OSError
```

Comment:
I agree. I picked this mainly due to existing usage in block io and to try and limit the number of changes (more on that below).

The previous code did not raise any error when attempts to read a block index failed: https://github.com/asdf-format/asdf/blob/b0c9a50e8d3add337e1271369a3bf834925176a4/asdf/block.py#L409

I think raising an error makes the most sense. The only usage of `read_block_index` is in `asdf._block.reader.read_blocks`: https://github.com/braingram/asdf/blob/78cd8185677aa1da795de3393d78b1645500e924/asdf/_block/reader.py#L222 where a failed attempt to read the block index will result in falling back to reading the blocks serially.

Using a custom exception seems like an improvement but the use of custom exceptions was not common in the old block code. I did not attempt to update this (as part of this PR) to use custom exceptions and instead tried to keep the errors consistent (both in type and message). In this situation (where there was no previous error) I opted to stick with the commonly used error type (which was originally IOError which python replaced with OSError) instead of implementing a custom exception.

Do you think a custom exception is called for here? If so, should it be public (perhaps simply adding it to `asdf.exceptions`)?

Reply:
I converted this to a custom private exception `BlockIndexError` (which does not inherit from `OSError`).
642fd7f1c244d9d7662a5e5563e2a98f6f30e358
Looking through the backlog of issues it looks like we already have one for applying this throughout asdf:
https://github.com/asdf-format/asdf/issues/1222